[MRG+1] FIX Pipelined fitting of Clustering algorithms, scoring of K-Means in pipelines

amueller opened the pull request:
Fixes #4063.

Also fixes the embarrassing bug where clustering algorithms can't be used in pipelines.

jnothman commented:
Don't be sad. As Gaël likes to say: it's not version 1 yet, stuff's allowed
to break and change :)

I can't say I have the answer here, yet, but I think you're thinking about
it the right way. Do we really require all estimators to take 2 positional
args to fit, then?

On 8 January 2015 at 13:35, Andreas Mueller <notifications@github.com>
wrote:

>  For the pipeline, it could be that previous steps in the pipeline use y
> so that it is not None, though. So you could do LDA, and then do
> clustering.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4064#issuecomment-69127780>
> .
>

amueller commented:
Yeah, I think this is the only way to make pipeline work as expected. Testing ``y is None`` is not good as in my example. The other option would be to use introspection or ``try`` but that is not the approach we have taken so far.

amueller commented:
reviews on this one would be nice as it makes testing somewhat simpler.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -252,6 +241,21 @@ def _check_transformer(name, Transformer, X, y):
+    y = (X[:, 0] * 4).astype(np.int)
+    y = multioutput_estimator_convert_y_2d(name, y)
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
```
You need to add a `warnings.simplefilter('always')` here so that the registry doesn't catch some warnings and create heisenbugs... Better, decorate the whole test with `@ignore_warnings`

amueller replied:
that is not particular to this test, though, right? and I'm not sure the whole tests should be decorated. These are more for the deprecation warnings.

jnothman replied:
I think it should be reasonable to ignore all `sklearn` warnings, at least, in common tests. Perhaps we should change `ignore_warnings` to only apply to `sklearn` modules.

amueller replied:
Interestingly enough, if I put ``@ignore_warnings`` on the test (not the check function) not test is run :-/

amueller replied:
I thought about adding a test for ``partial_fit``, but I'm not sure about it. The pipelines need to look quite a bit different to support ``partial_fit``, and there is no pressing reason currently to enforce taking a ``y`` in the API....

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -252,6 +241,21 @@ def _check_transformer(name, Transformer, X, y):
+        estimator = Estimator()
+    set_fast_parameters(estimator)
+    set_random_state(estimator)
+    estimator.fit(X, y)
```
Hmm. We're checking here whether the method takes more than one positional argument. Do we rather need a whitelist of argument names for the second arg so that we're not inadvertently setting `sample_weight`?

Also, do we need a parallel check for `partial_fit`, even if not currently used in `Pipeline`? Certainly it would simplify [this code](https://github.com/scikit-learn/scikit-learn/pull/3907/files#diff-a95fe0e40350c536a5e303e87ac979c4R386)

amueller replied:
I think the second positional argument should always be ``y``. This is the case for all estimators but MDS #4132. At least in master. is the ``sample_weights`` before ``y`` in the clustering algorithms?
That would complicate things, and also break the current pipelining code.

amueller replied:
The ``sample_weights`` in ``DBSCAN`` mess with this. I'm pretty sure we need to change something. After your PR,
``make_pipeline(SelectKBest(), DBSCAN()).fit(X, y)`` will use ``y`` as sample weights in DBSCAN.

amueller replied:
``fit_predict`` in DBSCAN has a different signature than ``fit`` ... and it is indeed still broken.
So, we need to test that unsupervised algorithms ignore ``y``? Or do we want to make pipeline use a named argument ``y``?

amueller replied:
We could check that ``y`` is ignored in the clustering algorithms (checking if ``labels_`` exist or for ``ClusterMixin``) but that doe

amueller commented:
This one will simplify future testing such as #4162 and #4136 by the way....

jnothman commented:
Firstly, doc/developers/index.rst needs fixing. Secondly, I think this change should encompass `partial_fit`: RBM, LSHForest, and IPCA are current offenders.

amueller commented:
What is the benefit in ``partial_fit`` apart from making the tests easier to write?

jnothman commented:
> What is the benefit in partial_fit apart from making the tests easier to write?

Consistency with `fit`, for one; the notion that Gaël [expressed](https://github.com/scikit-learn/scikit-learn/issues/4143#issuecomment-71623140) elsewhere as a polymorphic design principle; the ability to remove an if-else clause in `[sklearn.learning_curve](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/learning_curve.py#L217)`; and anticipation of meta-estimators employing `partial_fit` in the future or in external projects where the developer may otherwise forget the unsupervised case...

amueller commented:
Fair enough.
Where would you put this in the developers doc? The current ``y`` argument of Transformers is not documented anywhere as far as I can see :-/

amueller commented:
Actually, there was a comment about ``fit`` in there, something that was not tested / enforced. I added comments about ``transform`` and ``score`` but I'm not sure that is the best place.

jnothman commented:
I meant there are mentions of fit being implemented with one positional
arg. Should be removed.

On 29 January 2015 at 02:53, Andreas Mueller <notifications@github.com>
wrote:

>  Actually, there was a comment about fit in there, something that was not
> tested / enforced. I added comments about transform and score but I'm not
> sure that is the best place.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4064#issuecomment-71858062>
> .
>

amueller commented:
Where?

jnothman commented:
Oh, I didn't look at it closely enough. It actually gives an invocation
spec, not a def signature. But maybe it should:
http://scikit-learn.org/dev/developers/index.html#different-objects

On 29 January 2015 at 10:32, Andreas Mueller <notifications@github.com>
wrote:

>  Where?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4064#issuecomment-71939659>
> .
>

jnothman commented:
Obscurely, it shows 1-ary `score` as an example.

On 29 January 2015 at 14:05, Joel Nothman <joel.nothman@sydney.edu.au>
wrote:

> Oh, I didn't look at it closely enough. It actually gives an invocation
> spec, not a def signature. But maybe it should:
> http://scikit-learn.org/dev/developers/index.html#different-objects
>
> On 29 January 2015 at 10:32, Andreas Mueller <notifications@github.com>
> wrote:
>
>>  Where?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/4064#issuecomment-71939659>
>> .
>>
>
>

jnothman commented on doc/developers/index.rst:
```diff
@@ -716,8 +716,11 @@ is not met, an exception of type ``ValueError`` should be raised.
 make it possible to use the estimator as part of a pipeline that can
 mix both supervised and unsupervised transformers, even unsupervised
-estimators are kindly asked to accept a ``y=None`` keyword argument in
+estimators need to accept a ``y=None`` keyword argument in
```
Indeed, this doc was already here. I guess I only looked at a glance at what I thought was an issue, but not really.

jnothman commented:
Great. This LGTM apart from your test failure, which I take it will be fixed before the next reviewer gives a +1 ;)

amueller commented:
Thanks for the review. Damn, I thought I just fixed travis. Will look at it again.

amueller commented:
Fixed.

ogrisel commented:
As the motivation of this change was to make it possible to call the clustering models in pipelines I would like to integrate a non-regression "integration" test in the spirit of https://github.com/scikit-learn/scikit-learn/pull/4064#issuecomment-69127682 in `test_common`. Note: `clf` is a bad name for such as step, `model` or `estimator` would be a better name. Or even, just use `make_pipeline` for that test.

ogrisel commented:
Other than that +1 on my side.

amueller commented:
Hum, I totally wanted to do that integration test with pipelines, seems I forgot about it. Not sure what should be tested, probably "everything". This PR has tests for the "API definition' of Pipeline, not doing the integration.
I'll try that tomorrow.

ogrisel commented:
I was thinking of a bunch of sanity checks such as:

```
for est in all_classifiers():
   p = make_pipeline(est()).fit(X, y)
   e = est().fit(X, y)
   assert_array_equal(e.score(X, y), p.score(X, y))
```
and the similar clustering and regressors and maybe others as long as they implement `fit` and `score`.

amueller commented:
I thought about which functions we want to test, I guess that this issue only exists for ``fit`` and ``score`` so your tests should be fine (plus a ``if hasattr(est, "score")``. I thought about whether we want to test ``decision_function``, ``predict_proba`` and ``predict``, too.

amueller commented:
at least ``transform`` should also be checked

amueller commented:
added integration test, rebased. And now I'm not convinced that this is the right fix any more :-/ Maybe not passing ``y`` if it is ``None`` would be the better fix? And removing the ``y=None`` everywhere?
I am a bit confused.
I just did
``> git grep "transform(self, X, y=None)"```
which yields
```
sklearn/cluster/birch.py:    def transform(self, X, y=None):
sklearn/cluster/k_means_.py:    def fit_transform(self, X, y=None):
sklearn/cluster/k_means_.py:    def transform(self, X, y=None):
sklearn/decomposition/base.py:    def transform(self, X, y=None):
sklearn/decomposition/base.py:    def inverse_transform(self, X, y=None):
sklearn/decomposition/dict_learning.py:    def transform(self, X, y=None):
sklearn/decomposition/fastica_.py:    def fit_transform(self, X, y=None):
sklearn/decomposition/nmf.py:    def fit_transform(self, X, y=None):
sklearn/decomposition/pca.py:    def fit_transform(self, X, y=None):
sklearn/decomposition/pca.py:    def transform(self, X, y=None):
sklearn/decomposition/pca.py:    def fit_transform(self, X, y=None):
sklearn/decomposition/pca.py:    def inverse_transform(self, X, y=None):
sklearn/decomposition/truncated_svd.py:    def fit_transform(self, X, y=None):
sklearn/feature_extraction/dict_vectorizer.py:    def fit_transform(self, X, y=None):
sklearn/feature_extraction/dict_vectorizer.py:    def transform(self, X, y=None):
sklearn/feature_extraction/text.py:    def transform(self, X, y=None):
sklearn/kernel_approximation.py:    def transform(self, X, y=None):
sklearn/kernel_approximation.py:    def transform(self, X, y=None):
sklearn/kernel_approximation.py:    def transform(self, X, y=None):
sklearn/manifold/isomap.py:    def fit_transform(self, X, y=None):
sklearn/manifold/locally_linear.py:    def fit_transform(self, X, y=None):
sklearn/manifold/spectral_embedding_.py:    def fit_transform(self, X, y=None):
sklearn/neighbors/approximate.py:    def fit_transform(self, X, y=None):
sklearn/neighbors/approximate.py:    def transform(self, X, y=None):
sklearn/preprocessing/data.py:    def transform(self, X, y=None):
sklearn/preprocessing/data.py:    def fit_transform(self, X, y=None):
sklearn/random_projection.py:    def transform(self, X, y=None):
sklearn/tests/test_pipeline.py:    def transform(self, X, y=None):
```
That y is never used anywhere and not required by any API, is it?

amueller committed:
Add test that score takes y, fix KMeans, FIX pipeline compatibility of clustering algorithms!

amueller commented:
Anyhow, this PR is good to go, I think. I removed the requirement on ``transform`` (which is not a requirement at all) from the developer docs.

ogrisel commented:
Great, merging this PR. Thanks again @amueller !

amueller commented:
Thanks for the reviews @ogrisel and @jnothman :)

jnothman commented:
> Edit: this is the right fix, just not enough coffee.

:)

Rest assured this fix is in accordance with what's been in the docs for some time...
